server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -22,11 +22,21 @@
 import java.util.concurrent.atomic.AtomicLong;
 
 /** Maps _uid value to its version information. */
-final class LiveVersionMap implements ReferenceManager.RefreshListener, Accountable {
+public final class LiveVersionMap implements ReferenceManager.RefreshListener, Accountable {
```

Comment:
I think we also need to ask the archiver if it is "unsafe" and consider the entire version map unsafe for GETs in that case? This can be done in a follow-up ofc if that was the intention. I think it is correct to not let `isUnsafe` do so, since that should still concern the refreshed state.

Reply:
The archiver currently doesn't have a notion of safe/unsafe. It only receives old maps which are evacuated current maps which contain entries that are inserted after we marked the map for safe access. My understanding is that we do not care about ids or archiving them when we are in the unsafe mode.
(This area is a bit tricky so I could be very wrong!).

Reply:
Although I think what I have now is wrong! Let's say:

1. [safe mode] operation **A** for id 3   --> at some point a refresh adds **A** to the archive
2. ... refreshes swap maps and we move to unsafe mode
3. [unsafe mode] operation **B** for id 3. --> since in unsafe mode nothing moves through the maps to the archiver
4. ... a switch to safe mode causes a refresh(in stateless also flush). archiver may or may not have operation **A**
5. get(id:3) comes in and we might return operation **A** from the archive, which is wrong.

I guess this is the case you meant? Once we move to unsafe we should mark archiver as unsafe (and clear it up since it will not be used anymore). The get in step 5 would find nothing, the `TransportGetFromTranslogAction` would return the current generation, which has been advanced since the `refresh("unsafe_version_map")` is also a flush (in the real-time get PR), the search shard would have to wait for the generation to serve the get. I guess this case of waiting is not avoidable since operation **B** was done when we were not tracking IDs (LiveVersionMap was unsafe)?

Reply:
I think that case would principally not need to know that the archive is safe, only the other way around is a problem. If we were to look up id "3" in the archive, it would be OK. Since if any non-optimized-adds had come in after "3" were indexed, we'd revert to a safe map and thus we do know that any *updates* will be reflected (assuming our uuid is collision free, which we assume).

However, the case I were thinking of is where we add "A" in unsafe mode, then transition to safe mode (twice, i.e., an extra refresh too). "A" is then not in the archive. A real-time GET will not see it neither in live version map or the archive and it will assume the search shard is up to date with respect to GET on "A" - and simply return to serve the request from the search shard.

Reply:
> then transition to safe mode (twice, i.e., an extra refresh too). "A" is then not in the archive.

But that transition [would flush](https://github.com/elastic/elasticsearch-serverless/pull/165/files#diff-5cf366f6709355345d07b54c95f1552ed837106c829685a2b061382057cb38bfR61). The `TransportGetFromTranslogAction` would return empty (as you said) but also includes its current generation which the search shard should wait for before handling the get request locally.

So if the case I mentioned is not a concern, the one you mentioned also seems to be handled by the real-time get PR, which means we don't need to care about tracking unsafe mode with the archive?

Reply:
That was a nice conclusion, thanks for reminding me. I think this deserves a well-crafted comment though (perhaps in the stateless code more than here).

Reply:
Thank you. I will. Although trying to write a test for the case i mentioned, I found out it could happen. Took me a while, because I found a mistake in the code here that would cause that case to pass. I've pushed https://github.com/elastic/elasticsearch/pull/95331/commits/8629e4754587c8d927ed70d1e7b945abdbf80955 to fix the issue that was masking that corner case. Now, a test like this would fail (which is the one I described above):
```java
public void testSafeUnsafeSafe() throws Exception {
    AtomicLong generation = new AtomicLong(0);
    var refresher = new TestRefresher();
    var statelessArchiver = new StatelessLiveVersionMapArchiver(generation::get, refresher);
    var map = new LiveVersionMap(statelessArchiver);
    assertFalse(map.isSafeAccessRequired());
    map.enforceSafeAccess();
    assertTrue(map.isSafeAccessRequired());
    var value1 = new IndexVersionValue(randomTranslogLocation(), 1, 1, 1);
    maybePutIndexUnderLock(map, "test1", value1);
    map.beforeRefresh();
    map.afterRefresh(randomBoolean());
    var value2 = new IndexVersionValue(randomTranslogLocation(), 2, 2, 1);  // -> this is just to move to unsafe
    maybePutIndexUnderLock(map, "test2", value2);
    map.beforeRefresh();
    map.afterRefresh(randomBoolean());
    assertFalse(map.isSafeAccessRequired());
    var value3 = new IndexVersionValue(randomTranslogLocation(), 3, 3, 1);
    maybePutIndexUnderLock(map, "test1", value3);
    map.beforeRefresh();
    map.afterRefresh(randomBoolean());
    assertEqualsUnderLock(map, "test1", value3);  //-> fails since it returns value1
}
```
Although that test is for the Serverless PR.

Reply:
This test above is I think not valid. (Only in that constructed test it behaves that way which is not how engine does the calls).

Reply:
I returned to this:

> But that transition [would flush](https://github.com/elastic/elasticsearch-serverless/pull/165/files#diff-5cf366f6709355345d07b54c95f1552ed837106c829685a2b061382057cb38bfR61).

I think we call
`refresh("unsafe_version_map", SearcherScope.INTERNAL, true);`
in that transition, which is not overridden to flush in stateless. I think we need to fix that (and to ensure we have a test provoking this). If I am right ofc.

Reply:
@henningandersen Well, the change that will cause a flush in that case, is in the real-time GET PR, which I think is where this would be used. Do you think that is necessary here? I can take out that part (two new refresh sources that should flush) out of the real-time GET PR and merge it separately.

Reply:
That is fine by me, whatever is more logical for you. Thanks for clarifying.